osf/management/commands/import_EGAP.py

```diff
@@ -0,0 +1,142 @@
+from website.settings import WATERBUTLER_INTERNAL_URL
+from osf_tests.factories import ApiOAuth2PersonalTokenFactory
+from framework.auth.core import Auth
+
+logger = logging.getLogger(__name__)
+HERE = os.path.dirname(os.path.abspath(__file__))
+
+
+def ensure_egap_schema():
+    schema = ensure_schema_structure(from_json('egap-registration.json'))
+    schema_obj, created = RegistrationSchema.objects.update_or_create(
+        name=schema['name'],
+        schema_version=schema.get('version', 1),
+        defaults={
+            'schema': schema,
+        }
+    )
+    if created:
+        schema_obj.save()
+
+
+def get_creator_auth_header(creator_username):
+    creator = OSFUser.objects.get(username=creator_username)
+    token = ApiOAuth2PersonalTokenFactory(owner=creator)
+    token.save()
+    return creator, {'Authorization': 'Bearer {}'.format(token.token_id)}
+
+
+def create_node_from_project_json(egap_assets_path, epag_project_dir, creator):
+    with open(os.path.join(egap_assets_path, epag_project_dir, 'project.json'), 'r') as fp:
+        project_data = json.load(fp)
+        title = project_data['title']
+        node = Node(title=title, creator=creator)
+        node.save()  # must save before adding contribs for auth reasons
+
+        for contributor in project_data['contributors']:
+            node.add_contributor_registered_or_not(Auth(creator), full_name=contributor['name'], email=contributor['email'])
+
+        node.set_visible(creator, visible=False, log=False, save=True)
+
+    return node
+
+
+def recursive_upload(auth, node, dir_path, parent='', metadata=list()):
+    for item in os.listdir(dir_path):
+        item_path = os.path.join(dir_path, item)
+        base_url = '{}/v1/resources/{}/providers/osfstorage/{}'.format(WATERBUTLER_INTERNAL_URL, node._id, parent)
+
+        if os.path.isfile(item_path):
+            with open(item_path, 'rb') as fp:
+                url = base_url + '?name={}&kind=file'.format(item)
+                resp = requests.put(url, data=fp.read(), headers=auth)
+
+            if resp.status_code != 201:
+                raise Exception('Error waterbutler response is {}, with {}'.format(resp.status_code, resp.content))
+
+            metadata.append(resp.json())
+        else:
+            url = base_url + '?name={}&kind=folder'.format(item)
+            resp = requests.put(url, headers=auth)
+
+            if resp.status_code != 201:
+                raise Exception('Error waterbutler response is {}, with {}'.format(resp.status_code, resp.content))
+
+            metadata.append(resp.json())
+
+            metadata = recursive_upload(auth, node, item_path, parent=resp.json()['data']['attributes']['path'], metadata=metadata)
+
+    return metadata
+
+
+def main(creator_username):
+    ensure_egap_schema()
+    creator, creator_auth = get_creator_auth_header(creator_username)
+
+    egap_assets_path = os.path.join(HERE, 'EGAP')
```

Comment:
I was talking with MattF, and he suggested that the easiest way to get it on the server was to create a private OSF project that we upload everything to, then provide that GUID to one of the Matts to run this command against.

As I'm thinking through, if we go this route, then it might be easiest to work just from a zipped file, rather than going through multiple directories, mainly to keep from having to recursive upload to the OSF. Which can be done, but is sometimes fraught with peril (unfortunately).

What are your thoughts on this? Both pulling from an OSF project and working from a zip file rather than a directory? I don't know if you can directory walk just inside the zip, but even unziiping into a tmp directory and working from there will be much easier than uploading 1500 directories to OSF.

Reply:
Yeah we can just download the zip for a private project and unzip it in local memory within the management command that's alright with me.